testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs

```diff
@@ -0,0 +1,517 @@
+            .map_err(|_| "Failed to sign StackerDB chunk")?;
+
+        let stackerdb_tx = stackerdbs.tx_begin(miners_db_config.clone()).map_err(|e| {
+            warn!("Failed to begin stackerdbs transaction to write .miners message"; "err" => ?e);
+            "Failed to begin StackerDBs transaction"
+        })?;
+
+        match stackerdb_tx.put_chunk(&miners_contract_id, chunk, event_dispatcher) {
+            Ok(()) => {
+                debug!("Wrote message to stackerdb: {message:?}");
+                Ok(())
+            }
+            Err(e) => {
+                warn!("Failed to write message to stackerdb {e:?}");
+                Err("Failed to write message to stackerdb".into())
+            }
+        }
+    }
+
+    pub fn begin_sign(
+        &mut self,
+        block: &NakamotoBlock,
+        block_attempt: u64,
+        burn_tip: &BlockSnapshot,
+        burnchain: &Burnchain,
+        sortdb: &SortitionDB,
+        stackerdbs: &mut StackerDBs,
+        event_dispatcher: &EventDispatcher,
+    ) -> Result<ThresholdSignature, NakamotoNodeError> {
+        let sign_id = Self::get_sign_id(burn_tip.block_height, burnchain);
+        let sign_iter_id = block_attempt;
+        let reward_cycle_id = burnchain
+            .block_height_to_reward_cycle(burn_tip.block_height)
+            .expect("FATAL: tried to initialize coordinator before first burn block height");
+        self.coordinator.current_sign_id = sign_id;
+        self.coordinator.current_sign_iter_id = sign_iter_id;
+
+        let block_bytes = block.serialize_to_vec();
+        let nonce_req_msg = self
+            .coordinator
+            .start_signing_round(&block_bytes, false, None)
+            .map_err(|e| {
+                NakamotoNodeError::SigningCoordinatorFailure(format!(
+                    "Failed to start signing round in FIRE coordinator: {e:?}"
+                ))
+            })?;
+        Self::send_signers_message(
+            &self.message_key,
+            sortdb,
+            burn_tip,
+            stackerdbs,
+            nonce_req_msg.into(),
+            self.is_mainnet,
+            &self.miners_db_config,
+            event_dispatcher,
+        )
+        .map_err(NakamotoNodeError::SigningCoordinatorFailure)?;
+
+        let Some(ref mut receiver) = self.receiver else {
+            return Err(NakamotoNodeError::SigningCoordinatorFailure(
+                "Failed to obtain the StackerDB event receiver".into(),
+            ));
+        };
+
+        let start_ts = Instant::now();
+        while start_ts.elapsed() <= self.signing_round_timeout {
+            let event = match receiver.recv_timeout(Duration::from_millis(50)) {
+                Ok(event) => event,
+                Err(std::sync::mpsc::RecvTimeoutError::Timeout) => {
+                    continue;
+                }
+                Err(std::sync::mpsc::RecvTimeoutError::Disconnected) => {
+                    return Err(NakamotoNodeError::SigningCoordinatorFailure(
```

Comment:
Is this recoverable? Or even reachable?

Reply:
It's not recoverable, and I don't think its reachable.

Reply:
It would now be reachable if a miner thread didn't clean up before another miner thread started. The prior miner thread could reach this error at which point it should exit.